Rebrand Diagnostics Server to DiagnosticPort

josalem opened the pull request:
closes #39792

_**PREVIEW BREAKING CHANGES**_
These changes will be breaking from preview 8.

These changes rebrand the customer facing portions of the Diagnostics Server to be called DiagnosticPort.  This is a simpler naming scheme than the previous usage of "forward" and "reverse" servers.

In addition, this solves one major issue of the configuration of DiagnosticPorts.  under the previous configuration knobs, users could only specify 1 "reverse" monitoring agent.  If anyone else attempted to configure a monitoring agent, they would either clobber the user's configuration or be clobbered by the user.  The new configuration knobs allow for multiple diagnostic ports to be configured in a `;` delimited list.

## Nomenclature and Syntax

A `DiagnosticPort` can be configured using the `DOTNET_DiagnosticPorts` environment variable.  The syntax for this configuration is as follows:

```
DOTNET_DiagnosticPorts=<port address>[,tag[...]][;<port address>[,tag[...]][...]]
```

Where:
* `<port address> ::= A NamedPipeName without \\.\pipe\ on Windows | the full path to Unix domain socket on other platforms`
* `tag ::= SUSPEND_MODE | PORT_TYPE`
* `SUSPEND_MODE ::= suspend | nosuspend` (default value is `nosuspend`)
* `PORT_TYPE ::= connect` (future types such as additional `listen` ports could be added to this list)

An example value would be:
```
DOTNET_DiagnosticPorts=$DOTNET_DiagnosticPorts;~/myawesome.sock,connect,nosuspend;/mounted/dir/alternatelonnectablec.sock
```

`DOTNET_DoagnosticPorts` supersedes `DOTNET_DiagnosticMonitorAddress` and `DOTNET_DiagnosticMonitorPauseOnStart` so the latter two have been removed with these changes.

Any Diagnostic Ports specified in this configuration are _in addition to_ the default port, i.e., `dotnet-diagnostic-<pid>-epoch>`.  The suspend mode of the default port is set via the new environment variable `DOTNET_DefaultDotnetPortSuspend` which defaults to `0` for `nosuspend`.

Each port configuration specifies whether it is a `suspend` or `nosuspend` port.  `suspend` induces the PauseOnStart functionality and causes the runtime to halt progress in `EEStartupHelper` before most usbsystems are online.  Since multiple ports can individually request `suspend`, the resume command now needs to be sent **_by each `suspend` port connection_** before the event will be set and the runtime resumes execution.

If a config specifies multiple tag values from a tag type, e.g., `<path>,nosuspend,suspend,suspend`, the **_first_** one is the only one respected.

The `port address` value is required for a port config.  If a config doesn't specify an address and only specifies tags, e.g., `suspend`, then that value will be treated as the path.

The runtime will make a best attempt to generate a port from a port config.  A bad port config won't cause an error state, but could lead to consumed resources, e.g., causing the runtime to continuously poll for a connect port that will never exist.

### additional changes

* fix scoping issue for the runtime identifier GUID used in the process info command and advertise protocol.  Previously, it was marked as a `static` namespace-scoped variable causing each translation unit to get its copy of the variable.  Changing this to `extern` and intializing the value in `DiagnosticServer::Initialize` before use, solves this problem.  Fixes #40696.

* Add `EventPipe::Disable` deferral logic for window of time between resuming the runtime from Pause on Start state to when `EventPipe::FinishInitialization` is called.  A disable in this time frame can cause `EventPipe::StartStreaming` to be called on a session that is in the middle of being disabled.  `EventPipe::Disable` is safe before resuming and after `EventPipe::FinishInitialization`.  `EventPipe::Disable` now checks the `s_CanStartThreads` variable (value is `true` _after_ `FinishInitilization` is complete) under a lock to see whether it can disable or should defer.  The deferred disabled are then safely acted on in `FinishInitialization`

CC @tommcdon @jander-msft @dotnet/dotnet-diag

Commit:
Rebrand Diagnostic Server to Diagnostic Port
* allows for multiple ports to be specified
* multiple ports can be in a suspend state
* resume only works if all suspend ports have resumed

Commit:
Update tests to use new config switches
* small bug fixes in config parsing code

josalem commented:
@danmosemsft where can I find the details for a "breaking change doc"?  Does this still need one if the things that are being broken were only introduced in previous previews?

josalem commented:
/azp run runtime

Commit:
Update diagnosticserver.cpp

josalem commented:
/azp run runtime

Commit:
Appease gcc warning

Commit:
Merge branch 'dev/josalem/diagnostic-port' of github.com:josalem/runtime into dev/josalem/diagnostic-port

Commit:
Const shenanigans

Commit:
Add tests for DiagnosticPort functionality

Commit:
Attempting to appease GCC

jander-msft commented on src/coreclr/src/inc/clrconfigvalues.h:
```diff
@@ -710,10 +710,10 @@ RETAIL_CONFIG_DWORD_INFO(INTERNAL_EventPipeCircularMB, W("EventPipeCircularMB"),
-RETAIL_CONFIG_STRING_INFO_EX(EXTERNAL_DOTNET_DiagnosticsMonitorAddress, W("DOTNET_DiagnosticsMonitorAddress"), "NamedPipe path without '\\\\.\\pipe\\' on Windows; Full path of Unix Domain Socket on Linux/Unix.  Used for Diagnostics Monitoring Agents.", CLRConfig::DontPrependCOMPlus_);
-RETAIL_CONFIG_DWORD_INFO_EX(EXTERNAL_DOTNET_DiagnosticsMonitorPauseOnStart, W("DOTNET_DiagnosticsMonitorPauseOnStart"), 1, "If DOTNET_DiagnosticsMonitorAddress is set, this will cause the runtime to pause during startup.  Resume using the Diagnostics IPC ResumeStartup command.", CLRConfig::DontPrependCOMPlus_);
+RETAIL_CONFIG_DWORD_INFO_EX(EXTERNAL_DOTNET_DiagnosticPortSuspend, W("DOTNET_DiagnosticPortSuspend"), 0, "This will cause the runtime to pause during startup before major subsystems are started.  Resume using the Diagnostics IPC ResumeStartup command.", CLRConfig::DontPrependCOMPlus_);
+RETAIL_CONFIG_STRING_INFO_EX(EXTERNAL_DOTNET_DiagnosticPorts, W("DOTNET_DiagnosticPorts"), "A semicolon delimited list of additional Diagnostic Ports, where a Diagnostic Port is a NamedPipe path without '\\\\.\\pipe\\' on Windows or the full path of Unix Domain Socket on Linux/Unix followed by optional tags, e.g., '<path>,listen,nosuspend;<path>,connect'", CLRConfig::DontPrependCOMPlus_);
```
One problem that I've been thinking about is if the path portion of the port description is a full path, then there are likely to be collisions if multiple processes see the same port description (because this is provided by an environment variable). At best, there is a race to see which process can establish its endpoint first and all of the others will silently fail. I'm wondering if the path portion needs to be the "directory" of where to establish the pipe or socket file and then the runtime uses its default naming convention of the file itself.

josalem replied:
That's certainly possible for a `listen` port, but I'm not sure we would want users doing that.  That scenario was the impetus for reversing the pipe in the first place.  I Think it would be fair to simply document that if you specify a `listen` port, it should be specified for only a single process.  If you are in a multi-process scenario, I would expect users to only use `connect` ports.

jander-msft replied:
You don't necessarily have control of what the process spawns. Third party libraries could spawn their own .NET process and very likely don't know to disable diagnostics on them. Even "dotnet run" spawns multiple processes.

josalem replied:
I'm comfortable just not allowing `listen` ports to be specified other than the default one.

josalem replied:
I talked with @tommcdon offline, and we decided it is safer to just disable the ability to specify additional `listen` ports for now.  If later we decide to re-enable it, then we can add limitations or specifiers to the name to solve this problem.

jander-msft replied:
Just for clarification, that means we'll remove the "listen" and "connect" tags since all ports are connect mode?

josalem replied:
I'll leave the code for consuming it but ignore the tag if it's seen.  It'll still be legal to specify `connect`, I'll just ignore `listen` and make the default `connect`.

noahfalk replied:
The issue text up top appears inconsistent now btw, in some places it has been updated to be connect-only, in others it still suggests listen is supported.

josalem replied:
Updated the issue text.

Commit:
Defer disable till after streaming starts
* detect the case where a disable comes in
  immediately following a ResumeRuntime.
  We need to defer the disable to happen AFTER streaming starts since
  there is locking that happens
  in both FinishInitialization and
  Disable

josalem commented on src/coreclr/src/vm/eventpipe.cpp:
```diff
@@ -435,6 +450,27 @@ void EventPipe::Disable(EventPipeSessionID id)
     }
     CONTRACTL_END;
 
+    // EventPipe::Disable is called synchronously since the diagnostics server is
+    // single threaded.  HOWEVER, if the runtime was suspended in EEStartupHelper,
+    // then EventPipe::FinishInitialize might not have executed yet.  Disabling a session
+    // needs to either happen before we resume or after initialization.  We briefly take the
+    // lock to check s_CanStartThreads to check whether we've finished initialization.  We 
+    // also check whether we are still suspended in which case we can safely disable the session
+    // without deferral.
+    {
+        CrstHolder _crst(GetLock());
+        if (!s_CanStartThreads && !IpcStreamFactory::AnySuspendedPorts())
+        {
+            s_rgDeferredDisableEventPipeSessionIds.Push(id);
+            return;
+        }
+    }
+
+    DisableHelper(id);
```
@sywhang can you double check my logic here.  We can chat on Monday about this case.

sywhang replied:
I am OOF until Tuesday but I can take a look at the changes over the weekend

sywhang reviewed (commented):
Few stuff but looks like it's coming along nicely 👍

Commit:
Rename ConnectionMode enum values

josalem commented on src/coreclr/src/vm/ipcstreamfactory.cpp:
```diff
@@ -36,90 +53,161 @@ bool IpcStreamFactory::ClientConnectionState::GetIpcPollHandle(IpcStream::Diagno
+    if (dotnetDiagnosticPortsW != nullptr)
     {
-        if (pIpc->Listen(callback))
+        nCharactersWritten = WideCharToMultiByte(CP_UTF8, 0, dotnetDiagnosticPortsW, -1, NULL, 0, NULL, NULL);
```
I'm not sure that the conversion to multi-byte chars is always `sizeof(WCHAR) * wcslen(str)`, so we need to call this method to get the correct number of `char` to reserve.

Commit:
PR feedback

Commit:
Merge branch 'dev/josalem/diagnostic-port' of github.com:josalem/runtime into dev/josalem/diagnostic-port

Commit:
Merge branch 'master' into dev/josalem/diagnostic-port

Commit:
Fix #40696 and add test to cover that

Commit:
Merge branch 'dev/josalem/diagnostic-port' of github.com:josalem/runtime into dev/josalem/diagnostic-port

Commit:
Fix scope of runtime cookie

Commit:
Ignore listen ports for now

Commit:
default port type is connect

noahfalk commented on src/tests/tracing/eventpipe/diagnosticport/diagnosticport.cs:
```diff
@@ -0,0 +1,331 @@
+
+            return fSuccess;
+        }
+
```
We should probably add a test(s) that ensure the string parsing is robust. Stuff like:
0 length ports, invalid paths, invalid tags, multiple conflicting tags, not setting any tags to use defaults, a path whose name is identical to a tag name

josalem replied:
For conflicting tag, the logic will use that last tag value specified (which should be tested as you're suggesting).  Would expect conflicting or duplicate tags to produce an error?

josalem replied:
Added a test that tries various pathological and nonsensical values for the port configs.

noahfalk commented on src/tests/tracing/eventpipe/pauseonstart/pauseonstart.cs:
```diff
@@ -28,7 +28,7 @@ public static async Task<bool> TEST_RuntimeResumesExecutionWithCommand()
             Task<bool> subprocessTask = Utils.RunSubprocess(
                 currentAssembly: Assembly.GetExecutingAssembly(),
-                environment: new Dictionary<string,string> { { Utils.DiagnosticsMonitorAddressEnvKey, serverName } },
+                environment: new Dictionary<string,string> { { Utils.DiagnosticPortsEnvKey, $"{serverName},connect,suspend" } },
```
connect and suspend are the defaults and I'd guess our simple use-case tutorial won't include them. I'd recommend making the test mirror as closely as possible what we expect users to do which hopefully shouldn't make any difference here.

noahfalk commented on src/coreclr/src/vm/eventpipe.cpp:
```diff
@@ -435,6 +450,27 @@ void EventPipe::Disable(EventPipeSessionID id)
     }
     CONTRACTL_END;
 
+    // EventPipe::Disable is called synchronously since the diagnostics server is
```
This is fixing a separate pre-existing bug yes? I think its fine to include it for expediency sake but I recommend updating the issue/final commit message to call it out separately.

josalem replied:
Updated the PR text

noahfalk reviewed (approved):
LGTM! A few comments inline

josalem commented on src/coreclr/src/vm/diagnosticserver.cpp:
```diff
@@ -266,22 +258,20 @@ void DiagnosticServer::PauseForDiagnosticsMonitor()
-            }
+            CLRConfigStringHolder dotnetDiagnosticPortString = CLRConfig::GetConfigValue(CLRConfig::EXTERNAL_DOTNET_DiagnosticPorts);
+            WCHAR empty[] = W("");
+            wprintf(W("The runtime has been configured to pause during startup and is awaiting a Diagnostics IPC ResumeStartup command from a server in the following list:\n"));
```
Ah, fair point.  I'll amend the text to also print the default port suspend config variable.

josalem commented on src/coreclr/src/vm/ipcstreamfactory.h:
```diff
@@ -11,14 +11,67 @@
+        DiagnosticPortType Type = DiagnosticPortType::CONNECT;
+        DiagnosticPortSuspendMode SuspendMode = DiagnosticPortSuspendMode::NOSUSPEND;
+
+        DiagnosticPortBuilder WithPath(LPSTR path) { Path = path != nullptr ? _strdup(path) : nullptr; return *this; }
```
I believe it is copied once the port is built, so I should just wrap this in a holder or explicitly delete this pointer in the destructor.

josalem replied:
Trying to get too fancy with deletion led down a rabbit hole (yay rule of five and builder patter n in C++!), so I opted to just pass along a pointer into the parent string that is already in a holder that outlives the builder.

Commit:
PR feedback
* print default port's suspend status if waiting more than 5 sec
* remove strdup
* add test case for string parsing
* update tests to use defaults

josalem commented on src/tests/tracing/eventpipe/diagnosticport/diagnosticport.cs:
```diff
@@ -0,0 +1,331 @@
+
+                    Logger.logger.Log($"Stopping EventPipeSession");
+                    EventPipeClient.StopTracing(pid, sessionId);
+                    await readerTask;
```
Added an MRE instead of the bool flag.

josalem commented:
OSX Libraries failure appears to have been failing since at least [mid-July](https://dev.azure.com/dnceng/public/_build/results?buildId=728234&view=ms.vss-test-web.build-test-results-tab&runId=22544340&resultId=100085).  It looks like #30056.

josalem commented:
All green CI.  If there are no objections, I'll merge this soon.
